File: src/day19.zig
const std = @import("std");
const lib = @import("lib.zig");

pub const input = "data/input19.txt";

fn combinations(pattern: []const u8, towels: []const []const u8, alloc: std.mem.Allocator) !usize {
    var d = try alloc.alloc(usize, pattern.len + 1);
    defer alloc.free(d);
    @memset(d, 0);

    d[0] = 1;
    for (1..pattern.len + 1) |n| {
        for (towels) |t| {
            if (!std.mem.endsWith(u8, pattern[0..n], t)) continue;
            d[n] += d[n - t.len];
        }
    }

    return d[pattern.len];
}

pub fn solve(rdr: anytype, part: lib.Part) !usize {
    const alloc = std.heap.page_allocator;
    var lr = try lib.LineReader.init(rdr, alloc);
    defer lr.deinit();

    var towels = std.ArrayList([]const u8).init(alloc);
    defer towels.deinit();

    var iter = std.mem.splitSequence(u8, lr.lines[0], ", ");
    while (iter.next()) |towel| {
        try towels.append(towel);
    }

    var patterns = std.ArrayList([]const u8).init(alloc);
    defer patterns.deinit();
    for (lr.lines[2..]) |ln| {
        const s = std.mem.trim(u8, ln, &std.ascii.whitespace);
        try patterns.append(s);
    }

    var total: usize = 0;
    for (patterns.items) |pattern| {
        const n = try combinations(pattern, towels.items, alloc);
        total += if (part == .one) (if (n > 0) 1 else 0) else n;
    }
    return total;
}

pub fn main() !void {
    try lib.run(usize, solve, "Day 19", input);
}

test "test advent example" {
    const patterns: []const []const u8 = &.{
        "brwrr",
        "bggr",
        "gbbr",
        "rrbgbr",
        "ubwu",
        "bwurrg",
        "brgr",
        "bbrgwb",
    };
    const towels: []const []const u8 = &.{
        "r",
        "wr",
        "b",
        "g",
        "bwu",
        "rb",
        "gb",
        "br",
    };
    const expected = [_]usize{ 2, 1, 4, 6, 0, 1, 2, 0 };
    for (expected, 0..) |exp, n| {
        try std.testing.expect(try combinations(
            patterns[n],
            towels,
            std.testing.allocator,
        ) == exp);
    }
}

File: src/day19.py
import pytest
import sys

from collections import defaultdict


def find_combos_dp(pattern, towels):
    d = defaultdict(int)
    d[""] = 1  # empty is always possible for first iteration

    # dynamic programming table
    for i in range(1, len(pattern) + 1):
        partial_pattern = pattern[:i]
        for t in towels:
            # only consider towels that can be placed at end
            if not partial_pattern.endswith(t):
                continue

            # if we place this towel at the end, we're
            # basing it on the number of patterns we can
            # make with the pattern minus this towel
            # ie. d['abc'] = d['ab'] + towel_c
            last_chunk = partial_pattern[: -len(t)]
            d[partial_pattern] += d.get(last_chunk, 0)

    return d[pattern]


def is_possible(pattern, towels):
    """
    Original solution but ground to a halt for part two.
    So I searched for a dynamic programming solution.
    """
    partials = set([t for t in towels if pattern.startswith(t)])

    while partials:
        this_turn = set()
        for towel in towels:
            for partial in partials:
                candidate = partial + towel
                if pattern == candidate:
                    return True
                if pattern.startswith(candidate):
                    this_turn.add(candidate)
        partials = this_turn
    return False


def count_num_possible(patterns, towels):
    return sum(find_combos_dp(p, towels) > 0 for p in patterns)


def count_combinations(patterns, towels):
    return sum(find_combos_dp(p, towels) for p in patterns)


def main():
    with open("data/input19.txt") as f:
        data = f.read()

    towels = data.split("\n\n")[0].strip().split(", ")
    patterns = data.split("\n\n")[1].strip().split("\n")

    print(f"part 1: {count_num_possible(patterns, towels)}")
    print(f"part 2: {count_combinations(patterns, towels)}")


def test_example_part_one():
    patterns = ["brwrr", "bggr", "gbbr", "rrbgbr", "ubwu", "bwurrg", "brgr", "bbrgwb"]
    towels = ["r", "wr", "b", "g", "bwu", "rb", "gb", "br"]
    assert count_num_possible(patterns, towels) == 6


def test_example_part_two():
    patterns = ["brwrr", "bggr", "gbbr", "rrbgbr", "ubwu", "bwurrg", "brgr", "bbrgwb"]
    towels = ["r", "wr", "b", "g", "bwu", "rb", "gb", "br"]
    assert count_combinations(patterns, towels) == 16


if __name__ == "__main__":
    if len(sys.argv) > 1 and sys.argv[1] == "test":
        pytest.main(["-xvvs", __file__])
    else:
        main()
